src/careamics/careamist.py

```diff
@@ -580,76 +652,217 @@ def predict(
-                    pred_data=source_path,
-                    read_source_func=read_source_func,
-                    extension_filter=extension_filter,
-                    dataloader_params=dataloader_params,
-                )
-
-                # record datamodule
-                self.pred_datamodule = datamodule
-
-                return self.trainer.predict(
-                    model=self.model, datamodule=datamodule, ckpt_path=checkpoint
-                )
-
-            elif isinstance(source, np.ndarray):
-                # create datamodule
-                datamodule = CAREamicsPredictData(
-                    pred_config=prediction_config,
-                    pred_data=source,
-                    dataloader_params=dataloader_params,
-                )
-
-                # record datamodule
-                self.pred_datamodule = datamodule
+        # record datamodule
+        self.pred_datamodule = self._create_pred_datamodule(
+            source,
+            batch_size=batch_size,
+            tile_size=tile_size,
+            tile_overlap=tile_overlap,
+            axes=axes,
+            data_type=data_type,
+            tta_transforms=tta_transforms,
+            dataloader_params=dataloader_params,
+            read_source_func=read_source_func,
+            extension_filter=extension_filter,
+        )
+        return self.trainer.predict(
+            model=self.model, datamodule=self.pred_datamodule, ckpt_path=checkpoint
+        )
 
-                return self.trainer.predict(
-                    model=self.model, datamodule=datamodule, ckpt_path=checkpoint
-                )
+    def predict_to_disk(
+        self,
+        source: Union[CAREamicsPredictData, Path, str, np.ndarray],
+        *,
+        batch_size: int = 1,
+        tile_size: Optional[Tuple[int, ...]] = None,
+        tile_overlap: Tuple[int, ...] = (48, 48),
+        axes: Optional[str] = None,
+        data_type: Optional[Literal["array", "tiff", "custom"]] = None,
+        tta_transforms: bool = True,
+        dataloader_params: Optional[Dict] = None,
+        read_source_func: Optional[Callable] = None,
+        extension_filter: str = "",
+        checkpoint: Optional[Literal["best", "last"]] = None,
+        save_type: Literal["tiff", "custom"] = "tiff",
+        save_extension: Optional[str] = None,
+        save_predict_func: Optional[SavePredictFunc] = None,
+        save_predict_func_kwargs: Optional[Dict[str, Any]] = None,
+        predict_dir: str | Path = "predict",
+        force: bool = False,
+    ) -> Union[List[Path], Path]:
+        """ """
+        self.model.set_save_prediction_args(
+            save_predictions=True,
+            save_type=save_type,
+            save_extension=save_extension,
+            save_predict_func=save_predict_func,
+            save_predict_func_kwargs=save_predict_func_kwargs,
+            predict_dir=predict_dir,
+            force=force,
+        )
```

Comment:
So if I understand correctly:
- `CallBack` is saving data
- Saving settings live in the `LightningModule`
- Prediction call set the settings in the `LightningModule`

It's quite convoluted and I don't really like that the settings live in the module, since they really don't belong there...

## Alternative 1

Could `CAREamist` keep a reference of the `CallBack` and modify the settings directly there?

## Alternative 2

Could we make an internal `_generate_prediction` method that yields prediction by prediction (or batch of tiles per batch of tiles), and have all the stitching and saving logic happening in the `CAREamist` (or refactor somewhere else, but called there)?

Then we could have a function to run through the generator in order to save predictions, and one function to run through the generator in order to stitch and then save the predictions? (or however it makes sense to organize this).

Here we would go back into the not using Ligthning callback.

Reply:
Yes, I agree it's convoluted and gross.

> It's quite convoluted and I don't really like that the settings live in the module, since they really don't belong there...

The reason I put the settings in the module is because the `write_on_batch_end` hook gets passed the trainer and module as arguments, so the settings can be accessed from there, see below.
https://github.com/CAREamics/careamics/blob/d90aa0e8906fb942cc35f384be83bf42d0624d84/src/careamics/lightning_prediction_writer.py#L29-L49 
However your alternative 1 would definitely solve this problem, by just having the settings saved in the callback directly!

For alternative two this means we would not really be using the lightning `trainer.predict` at all (we would but not properly). This is what the prediction loop does. It would work but would mean we would not be able to use any of the potential benefits of Lightning down the line, e.g. distributed predictions etc.

Reply:
Let's try with the callback then!